avocado/query/utils.py

```diff
@@ -0,0 +1,138 @@
+        settings = connections.databases[db]
+
+        # Add new database entry into connections handler so when the query
+        # executes the new connection will be accessible.
+        connections.databases[temp_db] = settings
+
+    # Get the connection to the temporary database.
+    conn = connections[temp_db]
+
+    # Get the backend specific process ID for the query. Note, this opens the
+    # connection to the database.
+    pid = _get_backend_pid(conn)
+
+    # Put database and pid in centralized cache so multiple threads
+    # and/or processes can access it. This is left to expire and/or
+    # be overwritten.
+    cache.set(temp_db, (db, pid))
+
+    return conn
+
+
+def cancel_query(name):
+    "Cancels a query running on a named connection."
+    temp_db = TEMP_DB_ALIAS_PREFIX.format(name)
+
+    # Connection is no longer available. Assume there is nothing to do.
+    if temp_db not in connections._databases:
+        return
+
+    info = cache.get(temp_db)
+
+    if info is not None:
+        db, pid = info
+
+        # Cancel the query. A boolean is returned denoting whether
+        # it was canceled.
+        canceled = _cancel_query(name, db, pid)
+    else:
+        canceled = None
+
+    close_connection(name)
+
+    return canceled
+
+
+def close_connection(name):
+    "Closes a temporary connection by name and removes it from the handler."
+    temp_db = TEMP_DB_ALIAS_PREFIX.format(name)
+
+    # Remove connection from handler
+    if temp_db in connections._databases:
+        conn = connections[temp_db]
+        conn.close()
+
+        del connections._databases[temp_db]
+
+
+def _get_backend_pid(conn):
+    "Gets the backend specific process or query ID for a connection."
+    engine = conn.settings_dict['ENGINE']
+    conn.ensure_connection()
+
+    # SQLite does not return a process id, so we use 0 to be a non-None value
+    if engine == 'django.db.backends.sqlite3':
+        return 0
+
+    if engine == 'django.db.backends.postgresql_psycopg2':
+        return conn.connection.get_backend_pid()
+
+    if engine == 'django.db.backends.mysql':
+        return conn.connection.thread_id()
+
+    logger.warn('getting PIDs for {0} is not supported'.format(engine))
+
+
+def _conn_info(name):
+    temp_db = TEMP_DB_ALIAS_PREFIX.format(name)
+    return cache.get(temp_db)
+
+
+def _cancel_query(name, db, pid):
+    conn = connections[db]
+    engine = conn.settings_dict['ENGINE']
+
+    if engine == 'django.db.backends.postgresql_psycopg2':
+        c = conn.cursor()
+        c.execute('SELECT pg_cancel_backend(%s)', (pid,))
+        canceled, = c.fetchone()
+        return canceled
+
+    if engine == 'django.db.backends.mysql':
+        c = conn.cursor()
+
+        # Kills the query, but not the connection. Owners of the thread
+        # are allowed to kill their own queries.
+        c.execute('KILL QUERY %s', (pid,))
+
+        # TODO; any way to confirm it was actually killed?
```

Comment:
Could you use [os.kill(...)](https://docs.python.org/2/library/os.html#os.kill)? This would require Python 2.7 and above. You could send a signal of 0 which should just return an exit code indicating if the signal could be sent. If the signal could not be sent then the process is killed. If the signal was sent successfully, then the process was not killed in the execute line above.